crates/holochain_state/src/source_chain.rs

```diff
@@ -788,45 +790,74 @@ where
                         )",
                     });
                     sql.push_str(
-                        "
+                        format!("
                         )
                         AND
-                        (:entry_type IS NULL OR Action.entry_type = :entry_type)
+                        (:entry_type IS NULL OR Action.entry_type IN ({}))
                         AND
-                        (:action_type IS NULL OR Action.type = :action_type)
-                        ORDER BY Action.seq 
-                        ",
+                        (:action_type IS NULL OR Action.type IN ({}))
+                        ORDER BY Action.seq
+                        ", named_param_seq("entry_type", query.entry_type.as_ref().map_or(0, |t| t.len())), named_param_seq("action_type", query.action_type.as_ref().map_or(0, |t| t.len()))).as_str(),
                     );
-                    sql.push_str(if query.order_descending {" DESC"} else {" ASC"});
+                    sql.push_str(if query.order_descending { " DESC" } else { " ASC" });
                     let mut stmt = txn.prepare(&sql)?;
+
+                    let mut args: Vec<(String, Box<dyn rusqlite::ToSql>)> = Vec::with_capacity(6 + query.entry_type.as_ref().map_or(0, |t| t.len()) + query.action_type.as_ref().map_or(0, |t| t.len()));
+                    args.push((":author".to_string(), Box::new(author)));
+
+                    match &query.entry_type {
+                        None => {
+                            args.push((":entry_type".to_string(), Box::new(None::<EntryType>.as_sql())))
+                        }
+                        Some(types) => {
+                            // Value should not be 'Some' until it has at least one value
+                            args.push((":entry_type".to_string(), Box::new(types.get(0).unwrap().as_sql())));
+                            for i in 1..types.len() {
+                                args.push((format!(":entry_type_{}", i), Box::new(types.get(i).unwrap().as_sql())));
+                            }
+                        }
+                    }
+
+                    match &query.action_type {
+                        None => args.push((":action_type".to_string(), Box::new(None::<EntryType>.as_sql()))),
+                        Some(types) => {
+                            // Value should not be 'Some' until it has at least one value
+                            args.push((":action_type".to_string(), Box::new(types.get(0).as_ref().unwrap().as_sql())));
+                            for i in 1..types.len() {
+                                args.push((format!(":action_type_{}", i), Box::new(types.get(i).unwrap().as_sql())));
+                            }
+                        }
+                    }
+
+                    args.push((":range_start".to_string(), Box::new(match query.sequence_range {
```

Comment:
I was thinking you could reverse this and start by `match query.sequence_range`, but realize you can't because you need to write an argument for each parameter.

Reply:
I spend quite a bit of time staring at this wonder how it could be done more cleanly. The docs from Rusqlite https://docs.rs/rusqlite/latest/rusqlite/struct.ParamsFromIter.html suggest these kinds of constructions are largely left up to the user. Their internal types are also not that easy to extend. What I really want to express is "flatten this repeated parameter into the named parameters" but doing it directly is where I landed